tools/amd_build/pyHIPIFY/hipify-python.py

```diff
@@ -0,0 +1,946 @@
+
+        output_string = input_string.replace(function_body, stub)
+
+    # Add HIP Preprocessors.
+    elif replace_style == 2:
+        output_string = input_string.replace(
+            function_body,
+            "#if !defined(__HIP_PLATFORM_HCC__)\n%s\n#endif" % function_body)
+
+    # Add HIP Preprocessors.
+    elif replace_style == 3:
+        output_string = input_string.replace(
+            function_body,
+            "#if !defined(__HIP_DEVICE_COMPILE__)\n%s\n#endif" % function_body)
+
+    # Throw an exception at runtime.
+    elif replace_style == 4:
+        stub = "%s{\n%s;\n}" % (
+            function_string,
+            'throw std::runtime_error("The function %s is not implemented.")' %
+            function_string.replace("\n", " "))
+        output_string = input_string.replace(function_body, stub)
+
+    elif replace_style == 5:
+        stub = "%s{\n%s;\n}" % (
+            function_string,
+            'assert(0)')
+        output_string = input_string.replace(function_body, stub)
+
+    elif replace_style == 6:
+        stub = "%s{\n;\n}" % (function_string)
+        output_string = input_string.replace(function_body, stub)
+    return output_string
+
+
+def preprocessor(filepath, stats):
+    """ Executes the CUDA -> HIP conversion on the specified file. """
+    with openf(filepath, "r+") as fileobj:
+        output_source = fileobj.read()
+
+        # Perform type, method, constant replacements
+        for mapping in CUDA_TO_HIP_MAPPINGS:
+            for cuda_type, value in mapping.items():
+                # Extract relevant information
+                hip_type = value[0]
+                meta_data = value[1:]
+
+                if output_source.find(cuda_type) > -1:
+                    # Check if supported
+                    if constants.HIP_UNSUPPORTED in meta_data:
+                        stats["unsupported_calls"].append((cuda_type, filepath))
+
+                if cuda_type in output_source:
+                    output_source = re.sub(r'\b(%s)\b' % cuda_type, lambda x: hip_type, output_source)
+
+        # Perform Kernel Launch Replacements
+        output_source = processKernelLaunches(output_source, stats)
+
+        # Disable asserts
+        if not filepath.endswith("THCGeneral.h.in"):
+            output_source = disable_asserts(output_source)
+
+        # Overwrite file contents
+        fileobj.seek(0)
+        fileobj.write(output_source)
+        fileobj.truncate()
+        fileobj.flush()
+
+        # Flush to disk
+        os.fsync(fileobj)
+
+
+def file_specific_replacement(filepath, search_string, replace_string, strict = False):
+    with openf(filepath, "r+") as f:
+        contents = f.read()
+        if strict:
+            contents = re.sub(r'\b(%s)\b' % search_string, lambda x: replace_string, contents)
+        else:
+            contents = contents.replace(search_string, replace_string)
+        f.seek(0)
+        f.write(contents)
+        f.truncate()
+        f.flush()
```

Comment:
This (and the fsync below) is totally unnecessary, it will flush on file close.

Reply:
done